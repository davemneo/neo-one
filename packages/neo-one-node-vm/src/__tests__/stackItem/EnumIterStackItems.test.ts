import { AsyncIterableX } from '@reactivex/ix-es2015-cjs/asynciterable';
import BN from 'bn.js';
import { deserializeStackItem } from '../..';
import {
  BooleanStackItem,
  EnumeratorStackItem,
  IntegerStackItem,
  StackItemEnumerator,
  StackItemIterator,
} from '../../stackItem';

describe('Enumerator/Iterator Stack Item Types', () => {
  const testIterator: ReadonlyArray<{ readonly key: IntegerStackItem; readonly value: IntegerStackItem }> = [
    { key: new IntegerStackItem(new BN(0)), value: new IntegerStackItem(new BN(1)) },
  ];
  const testAsyncIterable = AsyncIterableX.from(testIterator);

  const stackItemEnumerator = new StackItemEnumerator(testAsyncIterable[Symbol.asyncIterator]());
  const stackItemIterator = new StackItemIterator(testAsyncIterable[Symbol.asyncIterator]());
  const enumeratorStackItem = new EnumeratorStackItem(stackItemEnumerator);

  test('StackItemEnumerator - mutableCurrent throws before .next() called', () => {
    expect(() => stackItemEnumerator.value()).toThrowError(
      'Current is not set. The enumerator has been fully consumed or has not been initialized',
    );
  });

  test('StackItemEnumerator - start of iterating done - false', () => {
    expect(stackItemEnumerator.done).toBeFalsy();
  });

  test('StackItemEnumerator - equals', () => {
    expect(stackItemEnumerator.equals(stackItemEnumerator)).toBeTruthy();
    expect(stackItemEnumerator.equals(true)).toBeFalsy();
  });

  test('StackItemEnumerator - next', async () => {
    expect(await stackItemEnumerator.next()).toBeTruthy(); // first .next() has a value.
    expect(await stackItemEnumerator.next()).toBeFalsy(); // sets this.mutableDone to true, returns false
    expect(await stackItemEnumerator.next()).toBeFalsy(); // returns !this.mutableDone
  });

  test('StackItemIterator - mutableCurrent throws before initialized', () => {
    expect(() => stackItemIterator.key()).toThrowError(
      'Current is not set. The iterator has been fully consumed or has not been initialized',
    );
  });

  test('EnumeratorStackItem - toStructuralKey', () => {
    expect(enumeratorStackItem.toStructuralKey()).toMatchSnapshot();
  });

  test('EnumeratorStackItem - equals', () => {
    expect(enumeratorStackItem.equals(enumeratorStackItem)).toBeTruthy();
    expect(enumeratorStackItem.equals(true)).toBeFalsy();
  });
});

describe('Deserialize Stack Item Tests', () => {
  const booleanStackItem = new BooleanStackItem(true);
  const integerStackItem = new IntegerStackItem(new BN(0));

  const serialBooleanItem = booleanStackItem.serialize();
  const serialIntegerItem = integerStackItem.serialize();

  test('Deserialize BooleanStackItem', () => {
    expect(deserializeStackItem(serialBooleanItem)).toEqual(booleanStackItem);
  });

  test('Deserialize IntegerStackItem', () => {
    expect(deserializeStackItem(serialIntegerItem)).toEqual(integerStackItem);
  });
});
