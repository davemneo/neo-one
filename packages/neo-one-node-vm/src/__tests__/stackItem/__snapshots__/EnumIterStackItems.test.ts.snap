// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enumerator/Iterator Stack Item Types EnumeratorStackItem - toStructuralKey 1`] = `"1"`;
